text/0000-unified_coroutines.md

```diff
@@ -0,0 +1,411 @@
+
+The design we propose, in which the generator arguments are mentioned only at the start of the generator most closely resembles what is hapenning. And the user can't make a mistake by not assigning to the argument bindings from the yield statement. Only drawback of this approach is, the 'magic'. Since the value of the `name` is magically changed after each `yield`. But we pose that this is very similar to a closure being 'magically' transformed into a generator if it contains a `yield` statement and as such is acceptable amount of 'magic' behavior for this feature.
+
+![magic](https://media2.giphy.com/media/12NUbkX6p4xOO4/giphy.gif)
+
+Nonetheless, the introduction of this implicit behavior will require additional cognitive load for new users when learning this feature. However, the behavior of Generators without arguments is unchanged, and therefore this change does not impose this cost upfront, making it possible to introduce the more complex behavior in progressively more complex examples.
+
+This change would result in following generator trait.
+
+```rust
+pub trait Generator<Args> {
+    type Yield;
+    type Return;
+    fn resume(self: Pin<&mut Self>, args: Args) -> GeneratorState<Self::Yield, Self::Return>;
+}
+```
+
+While the RFC does not deal with the lifetimes of the arguments, the similarity of the modified `Generator` trait with the existing `Fn*` traits suggests that rules which currently apply to closures will also apply to generators. [More info later](theoretical-basis).
+
+### Use cases:
+1. Futures generated by async-await. The current implementation of async futures requires the use of thread-local storage in order to  pass the `task::Context` argument into underlying futures. This imposes small, but not zero overhead, which would be removed by this RFC.
+
+2. Protocol state machines - When a user wants to implement a state machine in order to correctly represent a network protocol, 
+the ususal approach is to create a `State` enum, and upon every state machine transition `mem::replace`the current state with a default one, and perform a match, to possibly generate a new state, which is then stored into the place of original state.
+
+Example:
+Consider an implementation of following state machine:
+```
+     a                 b
+  +-----+           +----+
+  |     |           |    |
+  |  +--v--+  b  +-----+ |
+  |  |     +----->     | |
+  +--+  F  |     |  S  <-+
+     |     <-----+     |
+     +-----+  a  +-----+
+
+```
+How the similar state machines are implemented today:
+```rust
+enum State { Empty, First, Second }
+
+enum Event { A, B }
+
+fn machine(state: &mut State, event: Event) -> &'static str {
+    match (mem::replace(state, State::Empty), event) {
```

Comment:
Related to my last comment, you don't need to replace `state` here with `State::Empty`